srtp.go

```diff
@@ -44,36 +44,63 @@ func (c *Context) DecryptRTP(dst, encrypted []byte, header *rtp.Header) ([]byte,
 	return c.decryptRTP(dst, encrypted, header)
 }
 
-func (c *Context) encryptRTP(dst, decrypted []byte, header *rtp.Header) ([]byte, error) {
-	dst = allocateIfMismatch(dst, decrypted)
+// EncryptRTP Encrypts a RTP packet
+// This method modifies rtp.Header based on the contents of plaintext.
+// TODO This function should take (*rtp.Packet) instead of ([]byte, rtp.Header)
```

Comment:
We designed the API so that people who are just trying to decrypt packets can use an API that makes sense for them (resembling libsrtp). I don't know if this is a change we will make.

Reply:
Yeah, I see how you're trying to make generic by taking a byte slice. I do think it was a good goal but it has a performance impact.

Take this example:

```
raw, _ := packet.Marshal()
raw, _ = EncryptRTP(raw, raw, nil)
```

It will actually cause a reallocation, because `raw` won't have the capacity for the extra 10 bytes. I don't expect the caller of the function to realize that the function is not actually encrypting in place.

To avoid this, you have the use the shiny new `MarshalTo`. It's something that I would only expect `EncryptRTP` to do because it knows the auth size.

```
// Allocate 10 extra bytes for the auth tag.
// Otherwise it will cause a reallocation.
raw := make([]byte, 0, packet.MarshalSize() + 10)
raw, _ = packet.MarshalTo(raw)
raw, _ = EncryptRTP(raw, raw, nil)
```

We currently have to `Unmarshal` for header information, immediately after we just called `Marshal`. It's not a huge performance hit but it's something. Hell, I actually expected the point of passing in a `rtp.Header` was to avoid doing this `Unmarshal`, but nope.

I do agree that a 3rd party should be able to encrypt a `[]byte` slice. I also think that a 3rd party should be able to encrypt a `*rtp.Packet`. So one of the two has to `Marshal` or `Unmarshal` before calling `EncryptRTP`.

current:
```
// WriteStreamSRTP.WriteRTP and ones that deal with rtp.Packet
raw, _ = packet.Marshal()
raw, _ = EncryptRTP(raw, raw, nil) // REALLOCATION

// other libraries that deal with raw []byte slices
raw, _ = EncryptRTP(raw, raw, nil) // REALLOCATION
```

proposed:
```
// WriteStreamSRTP.WriteRTP and ones that deal with rtp.Packet
raw, _ = EncryptRTP(raw, packet)

// other libraries that deal with raw []byte slices
p, _ := rtp.UnmarshalPacket(raw)
raw, _ = EncryptRTP(raw, raw) // REALLOCATION
```

Truth be told, it's a pretty minor gripe that requires a lot of refactoring. Backwards compatibility makes more sense.